text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+
+### Errors or difficulty in configuration files
+
+Configuration files provide an opportunity for misconfiguration. A way to mitigate this would be to provide clear messaging and fail quickly when misconfiguration occurs.
+
+## Prior art and alternatives
+
+The working group looked to the OpenTelemetry Collector and OpenTelemetry Operator for inspiration and guidance.
+
+### Alternative schema languages
+
+In choosing to recommend JSON schema, the working group looked at the following options:
+
+* [Cue](https://cuelang.org/) - A promising simpler language to define a schema, the working group decided against CUE because:
+  * Tooling available for validating CUE files in languages outside of Go were limited.
+  * Familiarity and learning curve would create problems for both users and contributors of OpenTelemetry.
+* [Protobuf](https://developers.google.com/protocol-buffers) - With protobuf already used heavily in OpenTelemetry, the format was worth investigating as an option to define the schema. The working group decided against Protobuf because:
+  * Validation errors are the result of serlization errors which can be difficult to interpret.
+  * Limitations in the schema definition language result in poor ergonomics if type safety is to be retained.
+
+## Open questions
+
+### How to handle environment variable / file config overlap?
+
+How does file configuration interact with environment variable configuration when both are present?
+
+* Solution 1: Ignore environment configuration when file configuration is present. Log a warning to the user indicating that multiple configuration modes were detected, but use the file configuration as the source of truth.
+* Solution 2: Superimpose environment configuration on top of file configuration when both are present. One problem with this is that environment variable configuration doesn’t map to file configuration in an intuitive way. For example, OTEL_TRACES_EXPORTER defines a list of span exporters to be paired with a batch span processor configured by the OTEL_BSP_* variables. What do we do if the file config already contains one or more processors with an exporter specified in OTEL_TRACES_EXPORTER? Essentially, do we merge or append the environment variable configuration?
+
+### How to handle no-code vs programmatic configuration?
+
+How should the SDK be configured when both no-code configuration (either environment variable or file config) and programmatic configuration are present? NOTE: this question exists today with only the environment variable interface available.
+
+* Solution 1: Make it clear that interpretation of the environment shouldn’t be built into components. Instead, SDKs should have a component that explicitly interprets the environment and returns a configured instance of the SDK. This is how the java SDK works today and it nicely separates concerns.
```

Comment:
The API which is included allows users to implement these types of overrides / customizations, without getting bogged down with the details of exactly how overriding (i.e. merging configs) works or the exact set of use cases that are supported.

The `Parse(file) -> config` API parses a file and returns a config model. The `Configure(config)` accepts a config model and configures SDK components. 

The most typical use case would be to do something like `Configure(Parse(file))`, where the parsed model is passed directly to Configure. 

However, nothing is stopping users from doing all sorts customizations to the config model as needed in their environment. Example:
```
// Custom function to merge the second config model into the first
Config merge(Config first, Config second) { ... }

// Custom function to produce config model from environment variable scheme
Config getConfigFromEnv() { .. }

...

var config1 = Parse("/config1.yaml"); // Parse first config file
var config2 = Parse("/config2.yaml"); // Parse second config file
var configFromEnv = getConfigFromEnv(); // Get config from env

Configure(merge(config1, config2)); // Configure SDK with config2 taking priority over config1
Configure(merge(config2, config1)); // Configure SDK with config1 taking priority over config2
Configure(configFromEnv); // Configure SDK with configFromEnv
Configure(merge(config1, configFromEnv)); // Configure SDK with configFromEnv taking priority over config1
Configure(merge(configFromEnv, config1)); // Configure SDK with config1 taking priority over configFromEnv
Configure(merge(merge(configFromEnv, config1), config2)); // Configure SDK with config2 taking priority over config1 taking priority over configFromEnv
```

Let's leave the exact behavior of how to merge config models as unspecified. We can add it later if needed, but ideally, users would just define their own merge semantics since defining general rules that make everyone happy will be difficult.